crypto/cipher/aes_gcm_double.c

```diff
@@ -0,0 +1,688 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "aes_gcm_double.h"
+#include "alloc.h"
+#include "err.h" /* for srtp_debug */
+#include "crypto_types.h"
+#include "cipher_types.h"
+
+srtp_debug_module_t srtp_mod_aes_gcm_double = {
+    0,               /* debugging is off by default */
+    "aes gcm double" /* printable module name       */
+};
+
+/*
+ * For now, we only support a full-length auth tag
+ */
+#define GCM_IV_LEN 12
+#define GCM_AUTH_TAG_LEN 16
+
+
+
+#define SRTP_AES_128_DOUBLE_KEY_LEN                                     \
+    (SRTP_AES_128_KEY_LEN + SRTP_AES_128_KEY_LEN)
+#define SRTP_AES_256_DOUBLE_KEY_LEN                                     \
+    (SRTP_AES_256_KEY_LEN + SRTP_AES_256_KEY_LEN)
+
+// +-+-+-+-+-+-+-+-+
+// |R R R R B M P Q|
+// +-+-+-+-+-+-+-+-+
+#define OHB_MODIFIED_SEQ 0x01
+#define OHB_MODIFIED_PT  0x02
+#define OHB_MODIFIED_M   0x04
+
+#define RTP_HEADER_SIZE  12
+
+/*
+ * This function allocates a new instance of this crypto engine.
+ * The key_size parameter should be one of 56 or 88 for
+ * AES-128-GCM-DOUBLE or AES-256-GCM-DOUBLE respectively.  Note
+ * that the key length includes the 14 byte salt value that is
+ * used when initializing the KDF.
+ */
+static srtp_err_status_t srtp_aes_gcm_double_alloc(srtp_cipher_t **c,
+                                                int key_size,
+                                                int tlen)
+{
+    srtp_err_status_t err;
+    srtp_aes_gcm_double_ctx_t *dbl;
+
+    debug_print(srtp_mod_aes_gcm, "allocating cipher with key length %d",
+                key_size);
+    debug_print(srtp_mod_aes_gcm, "allocating cipher with tag length %d", tlen);
+
+    /*
+     * Verify the key_size is valid for one of: AES-128/256
+     */
+    if (key_size != SRTP_AES_GCM_128_DOUBLE_KEY_LEN_WSALT &&
+        key_size != SRTP_AES_GCM_256_DOUBLE_KEY_LEN_WSALT) {
+        return (srtp_err_status_bad_param);
+    }
+
+    if (tlen > GCM_DOUBLE_MAX_AUTH_TAG_LEN) {
+        return (srtp_err_status_bad_param);
+    }
+
+    /* allocate the base structs */
+    *c = (srtp_cipher_t *)srtp_crypto_alloc(sizeof(srtp_cipher_t));
+    if (*c == NULL) {
+        return (srtp_err_status_alloc_fail);
+    }
+    memset(*c, 0x0, sizeof(srtp_cipher_t));
+
```

Comment:
srtp_crypto_alloc returns memory that is already initialed to zero so the calls to memset are not required.